File: src/d1hashmaps/CS/DesignHashMaps.Algo/DesignHashMap.cs
// File: DesignHashMap.cs

namespace DesignHashMaps.Algo;

// DesignHashMap represents a hash map.
public class DesignHashMap
{
    private readonly int keySpace;
    private readonly Bucket[] buckets; // Array of Buckets

    public DesignHashMap()
    {
        keySpace = 2069;
        buckets = new Bucket[keySpace];

        for (int i = 0; i < keySpace; i++)
        {
            buckets[i] = new Bucket();
        }
    }

    // Private method to calculate the hash key
    private int GetHashKey(int key) => key % keySpace;

    // Put inserts a key-value pair into the hash map or updates the value if the key already exists.
    public void Put(int key, int value)
    {
        int hashKey = GetHashKey(key);

        buckets[hashKey].Update(key, value);
    }

    // Get returns the value associated with the key if it exists, otherwise returns -1.
    public int Get(int key)
    {
        int hashKey = GetHashKey(key);

        return buckets[hashKey].Get(key);
    }

    // Remove deletes the key-value pair from the hash map if it exists.
    public void Remove(int key)
    {
        int hashKey = GetHashKey(key);

        buckets[hashKey].Remove(key);
    }

    // PrintHashMap prints the current state of the hash map.
    public void PrintHashMap()
    {
        for (int i = 0; i < buckets.Length; i++)
        {
            if (buckets[i] != null && buckets[i].Count > 0)
            {
                Write($"Bucket {i}: ");
                buckets[i].PrintPairs();
                WriteLine();
            }
        }
    }
}

File: src/d1hashmaps/Java/DesignHashMap.java
import java.util.ArrayList;
import java.util.List;

public class DesignHashMap {
    private int keySpace;
    private List<Bucket> buckets;

    public DesignHashMap() {
        this.keySpace = 2069;
        this.buckets = new ArrayList<>();
        for (int i = 0; i < keySpace; i++) {
            this.buckets.add(new Bucket());
        }
    }

    public void put(int key, int value) {
        int hashKey = key % keySpace;
        buckets.get(hashKey).update(key, value);
    }

    public int get(int key) {
        int hashKey = key % keySpace;
        return buckets.get(hashKey).get(key);
    }

    public void remove(int key) {
        int hashKey = key % keySpace;
        buckets.get(hashKey).remove(key);
    }

    // For printing the current state of the hash map
    public void printHashMap() {
        for (int i = 0; i < buckets.size(); i++) {
            List<Pair> pairs = buckets.get(i).getPairs();
            if (pairs.size() > 0) {
                System.out.print("Bucket " + i + ": ");
                for (Pair pair : pairs) {
                    System.out.print("{key: " + pair.key + ", value: " + pair.value + "} ");
                }
                System.out.println();
            }
        }
    }
}
